pinot-core/src/main/java/org/apache/pinot/core/query/optimizer/filter/NumericalFilterOptimizer.java

```diff
@@ -448,21 +449,30 @@ private static void rewriteRangeOperator(Expression range, FilterKind kind, int
     }
   }
 
-  /** @return true if expression is a column of numeric type */
-  private static boolean isNumericColumn(Expression expression, Schema schema) {
-    if (expression.getType() != ExpressionType.IDENTIFIER) {
-      // Expression can not be a column.
-      return false;
-    }
-
-    String column = expression.getIdentifier().getName();
-    FieldSpec fieldSpec = schema.getFieldSpecFor(column);
-    if (fieldSpec == null || !fieldSpec.isSingleValueField()) {
-      // Expression can not be a column name.
-      return false;
+  /** @return field data type extracted from the expression. null if we can't determine the type. */
+  private static @Nullable FieldSpec.DataType getDataType(Expression expression, Schema schema) {
+    if (expression.getType() == ExpressionType.IDENTIFIER) {
+      String column = expression.getIdentifier().getName();
+      FieldSpec fieldSpec = schema.getFieldSpecFor(column);
+      if (fieldSpec != null && fieldSpec.isSingleValueField()) {
+        return fieldSpec.getDataType();
+      }
+    } else if (expression.getType() == ExpressionType.FUNCTION
+        && "cast".equalsIgnoreCase(expression.getFunctionCall().getOperator())) {
+      // expression is not identifier but we can also determine the data type.
+      String targetTypeLiteral = expression.getFunctionCall().getOperands().get(1).getLiteral().getStringValue()
+          .toUpperCase();
+      FieldSpec.DataType dataType;
+      if ("INTEGER".equals(targetTypeLiteral)) {
+        dataType = FieldSpec.DataType.INT;
+      } else if ("VARCHAR".equals(targetTypeLiteral)) {
+        dataType = FieldSpec.DataType.STRING;
+      } else {
+        dataType = FieldSpec.DataType.valueOf(targetTypeLiteral);
+      }
```

Comment:
Consider making a util for this part, and make `DataTypeConversionFunctions.cast()` and `CastTransformFunction` use the util. Currently they already have different behavior.

Reply:
they are not equivalent. conversion function operates `@ScalarFunction` which runs on `PinotDataType` (physical data type) and here it runs on `FieldSpec.DataType`.